DAS/das/utils/logger.py

```diff
@@ -0,0 +1,275 @@
+        """Graceful exit for closing the file and stopping the MQTT client."""
+        self._recording = False
+        self._CLIENT.loop_stop()
+        self._LOG_FILE.close()
+        logging.info(f"Data saved in {self._LOG_FILE.name}")
+
+
+class Playback:
+    """Playback MQTT messages from log files in realtime or faster.
+
+    Parameters
+    ----------
+    filepath : str
+        Filepath of the log file for playback
+    broker_address : str
+        The IP address that the MQTT broker lives on
+    verbose : bool
+        Specifies whether the outgoing MQTT data and warnings are printed
+
+    Attributes
+    ----------
+    _VERBOSE: bool
+        Specifies whether the incoming MQTT data and warnings are printed
+    _BROKER_ADDRESS: bool
+        The IP address of the MQTT broker
+    _log_data: list(dict)
+        A list of all of the rows in the log file stored in dict format
+    """
+
+    def __init__(
+        self, filepath: str, broker_address: str = "localhost", verbose: bool = False
+    ) -> None:
+        self._VERBOSE = verbose
+        self._BROKER_ADDRESS = broker_address
+
+        # Read in data from log file and save each row in _log_data list
+        log_file = open(filepath, "r")
+        csv_reader = csv.DictReader(
+            log_file,
+            delimiter=CsvConfig["delimiter"],
+            quotechar=CsvConfig["quotechar"],
+            quoting=CsvConfig["quoting"],
+            skipinitialspace=CsvConfig["skipinitialspace"],
+        )
+
+        self._log_data = []
+        for row in csv_reader:
+            # Convert time delta to float
+            row["time_delta"] = float(row["time_delta"])
+            self._log_data.append(row)
+
+    def play(self, speed: float = 1) -> None:
+        """Play the logged data at a certain speed using an async function.
+
+        Parameters
+        ----------
+        speed : float
+            Speed multiplier to determine how fast to send out the data. A higher value means faster.
+        """
+
+        logging.info(f"⚡ Playback initiated at {speed}x speed ⚡")
+
+        # Run the event loop to issue out all of the MQTT publishes
+        asyncio.run(self._publish(speed))
+
+    async def _publish(self, speed) -> None:
+        """Async function that collects all the necessary publish functions and gathers them to then be run by the
+        event loop.
+
+        Parameters
+        ----------
+        speed : float
+            Speed multiplier to determine how fast to send out the data. A higher value means faster.
+        """
+
+        async def _publish_aux(row):
+            scaled_sleep = row["time_delta"] / speed
+            await asyncio.sleep(scaled_sleep)
+
+            if self._VERBOSE:
+                logging.info(
+                    f"{round(row['time_delta'], 5): <10} | {round(scaled_sleep, 5): <10} | {row['mqtt_topic']: <50} | {row['message']}"
+                )
+
+            try:
+                publish.single(
```

Comment:
Kinda inclined to change the alias for `publish` to `mqtt_publish` or `publish_mqtt` (or something similar). I wasn't sure where this function came from since it's pretty similar to `_publish`.

Reply:
I've switched to using a client instead of just a single publish statement. This resolves this issue while also making the error messages much more clear!